### ISSUE TITLE
[Proposal] Edit inline with Entities Reference in choice field type.

### ISSUE THREAD

**Issue by Carteni** [sonata-project/SonataAdminBundle#4188]
Hi guys, congrats for your awesome work!
I'd like to inline-edit in list view injecting in the choice fields an entity id.

For example, in [HelperController::setObjectFieldValueAction method, before row 286](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/Controller/HelperController.php#L285):

```
// Check if value is an Entity Reference.
if ('' != $value && $fieldDescription->getOption('class')) {
    $value = $this->pool->getContainer()
     ->get('doctrine.orm.default_entity_manager')
     ->getRepository($fieldDescription->getOption('class'))
     ->find((int) $value);
}

$this->pool->getPropertyAccessor()->setValue($object, $propertyPath, '' !== $value ? $value : null);
```

This can be done with some tricks (overriding `sonata.admin.controller.admin` service, but it would be perfect if it were into the bundle.

Thanks!

**Comment by greg0ire** [comment]
I edited your message, you need to learn about triple backticks :wink:

**Comment by greg0ire** [comment]
> HelperController::setObjectFieldValueAction method, before row 286

Please edit your message to make this a link.

**Comment by greg0ire** [comment]
Hmmm… I think this piece of code could grow into a huge switch if we start trying everything we can to understand the client data. This sounds like an OCP violation, and I think there should be a better design than a bunch of ifs… maybe have an option to explicitely explain what is to be used instead of guessing it from the input? At least, the guessing part should be extracted from the controller.

**Comment by Carteni** [comment]
Maybe before ajax request arrives to HelperController::setObjectFieldValueAction method, we could convert `$request->get('value')` in Entity (for example catching the ajax request into a Listener and set the Entity into the value's request).

**Comment by greg0ire** [comment]
You mean with a `ParamConverter`?

**Comment by Carteni** [comment]
Something like this:

```
// app/config/services.yml
admin.param_converter.listener:
      class: AppBundle\EventListener\Admin\ParamConverterListener
      arguments: ['@doctrine.orm.entity_manager', '@sonata.admin.pool', '@security.http_utils']
      tags:
          - { name: kernel.event_listener, event: kernel.request, method: onRequest}


// AppBundle\EventListener\Admin\ParamConverterListener.php
public function onRequest(GetResponseEvent $event)
    {
        $request = $event->getRequest();

        if ($request->isXmlHttpRequest() &&
          $this->httpUtils->checkRequestPath(
            $request,
            'sonata_admin_set_object_field_value'
          )
        ) {

            $field = $request->get('field');
            $code = $request->get('code');
            $value = $request->get('value');

            $admin = $this->pool->getInstance($code);
            $fieldDescription = $admin->getListFieldDescription($field);

            if ('' != $value && $fieldDescription->getOption('class')) {
                $value = $this->em->getRepository(
                  $fieldDescription->getOption('class')
                )->find((int)$value);
                $request->request->set('value', $value);
            }
        }
    }

```

HelperController is untouched.

**Comment by greg0ire** [comment]
It's a good start but it would have to be more generic : you're searching in the ORM right now, but maybe you should be looking into the ODM (or any other persistence system supported by Sonata). IMO relying on a `class` option might not be enough

**Comment by Carteni** [comment]
You are right. Maybe with a new type field which extends 'choice' type?

```
protected function configureListFields(ListMapper $listMapper)
{
  $listMapper->add(
'category',
    'entity_choice', // instead choice
    array(
       'editable' => true,
       'choices' => $postCategoriesFlatten, // [1=>'Category 1', '2=>'Category 2','3=>'Category 3']
       'label' => 'Category'
    ),
    array(
       'object' => 'Application\Sonata\ClassificationBundle\Entity\Category',
       'manager' => 'doctrine.orm.entity_manager'
    )
)
}

**Comment by greg0ire** [comment]
I would something more like what happens in the form mapper : define the new field type in the doctrineormbundle, which means the name wouldn't be 'entity_choice', but 'doctrine_something', and you wouldn't need to specify the manager.

**Comment by Carteni** [comment]
Ok, and the ParamConverterListener in the doctrineormbundle too.

```
$listMapper->add(
'category',
    'doctrine_inline_edit' ???
```

**Comment by greg0ire** [comment]
Yes. Have a look at how the form mapper does this, after all, it's almost the same thing, right?

**Comment by Carteni** [comment]
But if i change 'choice' type field to 'doctrine_inline_edit' (fox eample), the column in ListMapper does remains editable or it converted in link?

**Comment by greg0ire** [comment]
I don't know, maybe it will have to be something separate like `inline_form_type` instead of just `type`

**Comment by jordisala1991** [comment]
This is already implemented #4449
